context_slog.go

```diff
@@ -0,0 +1,123 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2019 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"context"
+	"log/slog"
+)
+
+// This file contains the version of NewContext and FromContext which supports
```

Comment:
> I think I get what you are doing here, but I am not sure why.
>
> slog doesn't include "from context" functions - are you just "augmenting" their API by doing it here?

Yes, and because any interoperable API has to know about both (see below). That cannot be in slog.

> Do we really need this, or can we say "Go doesn't want to support it, go bother them" ?

That ship pretty much has sailed now. Suppose `slog.NewContext` and `slog.FromContext` were added now. Then consider this example:

```
func usingLogr(ctx context.Context) {
   logger := logr.FromContext(ctx)
   logger.Info("hello")
}

func usingSlog(ctx context.Context) {
    logger := slog.FromContext(ctx)
    logger.With("slogKey", "slogValue")
    ctx = slog.NewContext(ctx, logger)
    usingLogr(ctx)
}

func main() {
   logger := zapr.NewLogger(...)
   ctx := logr.NewContext(context.Background(), logger)
   usingSlog(ctx)
}
```

What we want here is that `"slogKey": "slogValue"` get added to the "hello" message. `logr.FromContext` can't achieve that.

It can look for a logger with its own key and call `slog.FromContext`. But the context has both and `slog.FromContext` cannot decide which one is more recent and thus should be used.

This PR avoids that by using the same key for all kinds of loggers.

Reply:
But as soon as slog realizes this was the wrong decision, we're back to this problem, right?

Reply:
My impression from the discussion was that the decision to not support passing via context is final.